### ISSUE TITLE
Rethink commands

### ISSUE THREAD

**Issue by jamesaoverton** [knocean/ontofetch#24]
Right now we have a `fetch` command that downloads a particular ontology file and its imports. We’re usually interested in an ontology project and its versions.

- `fetch dir url`: just fetch the URL to the directory; no checks, ignore the configuration file 
- `fetch project dir url`: just fetch the URL to the `project/dir` directory; no checks, ignore the configuration file
- `fetch project`: read the configuration file and fetch the ontology to a dated directory; no checks
- `status project`: read the configuration file and catalog, and determine whether a project has been updated; don’t download anything
- `status`: run status on all configured projects
- `update project`: run status, and if the project has been updated then `fetch project`
- `update`: run update on all projects

In other words, fetch should always download stuff. The status and update commands only make sense for a project, where we want to compare multiple versions.

**Comment by rctauber** [comment]
Note to self:
`extract` command to get owl:Ontology element, needs `-d` flag for where to put extracts.

**Comment by rctauber** [comment]
A few questions:

* `status` should just give notice if the project needs to be updated, correct? It should follow the redirects, get the ETag/Last-Modified, and if it's different, return a message. If this is included in `update`, do we need a separate command?
* All `fetch` commands - should these *just* fetch the URL, or the URL and all imports?

**Comment by jamesaoverton** [comment]
I've used this design before, and I think it makes sense. `update` is a conditional. `status` is the condition. `fetch` is the operation. So `update` means: if `status` is "update required" then `fetch`.

Yes, `status` should follow redirects, get headers, compare, and always return a message. It should not download anything, but if we don't have HTTP Headers or something we can trust then we don't know if an update is required. In that case we might have to download stuff. We'll start with the good case.

Yes, all `fetch` commands should also get imports.

Hmm.  If `status` is called and we don't have a previous version, then we immediately know that an update is required. If `status` is called and we have a previous version, then the catalog will contain the previous imports. I think we also need to check the status of all imports, because if any import has changed then the ontology has changed. As soon as any change is detected then we can return "update required" without further checks.

Rather than "has it changed" we really should be thinking in terms of "are we sure it has not changed?"

**Comment by rctauber** [comment]
> I've used this design before, and I think it makes sense. update is a conditional. status is the condition. fetch is the operation. So update means: if status is "update required" then fetch.

Sounds good

> Yes, status should follow redirects, get headers, compare, and always return a message. It should not download anything, but if we don't have HTTP Headers or something we can trust then we don't know if an update is required. In that case we might have to download stuff. We'll start with the good case.

Everything I've looked at so far has had an ETag, except anything that redirects to FTP. I'm guessing an FTP should always be a flag to re-fetch, since we can't be sure of it?

> Yes, all fetch commands should also get imports.

:+1:

> Hmm. If status is called and we don't have a previous version, then we immediately know that an update is required. If status is called and we have a previous version, then the catalog will contain the previous imports. I think we also need to check the status of all imports, because if any import has changed then the ontology has changed. As soon as any change is detected then we can return "update required" without further checks.

I think this goes back to [this comment](https://github.com/knocean/ontofetch/issues/16#issuecomment-350732736) for restructuring the way imports are cataloged. Is it just whatever imports have updated that need to be re-fetched, or does *everything* need to be re-fetched?

**Comment by jamesaoverton** [comment]
FTP doesn't have an equivalent of a HEAD request, so I think we always have to download.

If we running `fetch`, and there's a previous version and those files are still present, then I guess we can just fetch the files that have changed, and copy the files that have not changed from the previous directory/zipfile to the current directory/zipfile.

There should be a `fetch --force` option to force downloading new copies. If things work as expected, this should not be required, but you never know...

**Comment by rctauber** [comment]
So `fetch` would also run `status` first? Unless it includes `--force`, in which case it just gets everything.

**Comment by jamesaoverton** [comment]
Yes, `fetch` will require most of the information from `status`, which is most of the information in the catalog. We can discuss design details on a call if you like.

**Comment by rctauber** [comment]
Sounds good @jamesaoverton
